batch/plugins/org.jboss.tools.batch.core/src/org/jboss/tools/batch/internal/core/impl/BatchProject.java

```diff
@@ -0,0 +1,632 @@
+			artifacts.add(a);
+		}
+		synchronized (this) {
+			artifactsByPath.clear();
+			artifactsByName.clear();
+			artifactsByType.clear();
+			allArtifacts.clear();
+		}
+		for (IBatchArtifact a: artifacts) {
+			addArtifact(a);
+		}
+
+		if(updateDependent) {
+			Collection<BatchProject> dependent = new ArrayList<BatchProject>(usedBy);
+			for (BatchProject p: dependent) {
+				p.update(false);
+			}
+		}
+	}
+
+	public void addArtifact(IBatchArtifact artifact) {
+		addToMap(artifactsByName, artifact.getName(), artifact);
+		addToMap(artifactsByType, artifact.getArtifactType(), artifact);
+		addToMap(artifactsByPath, artifact.getSourcePath(), artifact);
+		synchronized (this) {
+			allArtifacts.add(artifact);
+		}
+	}
+
+	private synchronized <P> void addToMap(Map<P, Set<IBatchArtifact>> map, P key, IBatchArtifact artifact) {
+		if(key == null) {
+			return;
+		}
+		Set<IBatchArtifact> bs = map.get(key);
+		if(bs == null) {
+			bs = new HashSet<IBatchArtifact>();
+			map.put(key, bs);
+		}
+		bs.add(artifact);
+	}
+
+	public BatchProject[] getAllDependentProjects(boolean resolve) {
+		Map<BatchProject, Integer> set = new HashMap<BatchProject, Integer>();
+		getAllDependentProjects(set, 0);
+		if(resolve) {
+			for (BatchProject n: set.keySet()) {
+				n.resolve();
+			}
+		}
+		BatchProject[] result = set.keySet().toArray(new BatchProject[set.size()]);
+		Arrays.sort(result, new D(set));
+		return result;
+	}
+
+	public BatchProject[] getAllDependentProjects() {
+		return getAllDependentProjects(false);
+	}
+
+	private void getAllDependentProjects(Map<BatchProject, Integer> result, int level) {
+		if(level > 10) return;
+		BatchProject[] array = null;
+		synchronized(this) {
+			array = usedBy.toArray(new BatchProject[0]);
+		}
+		for (BatchProject n: array) {
+			if(!result.containsKey(n) || result.get(n).intValue() < level) {
+				result.put(n, level);
+				n.getAllDependentProjects(result, level + 1);
+			}
+		}
+	}
+	private static class D implements Comparator<BatchProject> {
+		Map<BatchProject, Integer> set;
+		D(Map<BatchProject, Integer> set) {
+			this.set = set;
+		}
+		@Override
+		public int compare(BatchProject o1, BatchProject o2) {
+			return set.get(o1).intValue() - set.get(o2).intValue();
+		}
+		
+	}
+
+	public DefinitionContext getDefinitions() {
+		return definitions;
+	}
+
+	public void pathRemoved(IPath path) {
+		definitions.getWorkingCopy().clean(path);
+	}
+
+	public void postponeFiring() {
+		//TODO
+	}
+
+	public void fireChanges() {
+		//TODO
+	}
+
+	public void reloadProjectDependencies() {
+		dependsOn.clear();
+		usedBy.clear();
+		loadProjectDependenciesFromKBProject();
+	}
+
+	private void loadProjectDependenciesFromKBProject() {
+		Element root = null;
+		File file = getKBStorageFile();
+		if(file != null && file.isFile()) {
+			root = XMLUtilities.getElement(file, null);
+			if(root != null) {
+				loadProjectDependencies(root);
+			}
+		}		
+	}
+	
+	private File getKBStorageFile() {
```

Comment:
Shouldn't we move all code which work with KBStorage and dependency elements to KB plugin?